python/activator/activator.py

```diff
@@ -195,6 +226,114 @@ def create_app():
+        "auto.offset.reset": fan_out_kafka_topic_offset,
+        "sasl.mechanism": fan_out_kafka_sasl_mechanism,
+        "security.protocol": fan_out_kafka_security_protocol,
+        "sasl.username": fan_out_kafka_sasl_username,
+        "sasl.password": fan_out_kafka_sasl_password,
+        'enable.auto.commit': False
+    }
+
+    fan_out_consumer = kafka.Consumer(fan_out_consumer_conf, logger=_log)
+
+    _log.info("starting fan out consumer")
+
+    try:
+        fan_out_consumer.subscribe([fan_out_kafka_topic])
+
+        fan_out_listen_start_time = time.time()
+
+        while time.time() - fan_out_listen_start_time < fanned_out_msg_listen_timeout:
+            fan_out_consumer.subscribe([fan_out_kafka_topic])
+            fan_out_message = fan_out_consumer.poll(timeout=5)
+            if fan_out_message is None:
+                continue
+            if fan_out_message.error():
+                if fan_out_message.error().code() == KafkaError._PARTITION_EOF:
+                    # End of partition event
+                    _log.debug('No more messages.  Reached end of offset')
+                if fan_out_message.error().code() == KafkaError._ALL_BROKERS_DOWN:
+                    _log.warning('All prompt processing kafka brokers are down')
+                if fan_out_message.error().code() == KafkaError._AUTHENTICATION:
+                    _log.error('Authentication failure with fanned out event topic')
+                    raise KafkaException(fan_out_message.error())
+                elif fan_out_message.error():
+                    _log.warning("Fanned out consumer event: %s", fan_out_message.error())
+            else:
+                deserialized_fan_out_visit = fan_out_avro_deserializer(fan_out_message.value(),
+                                                                       SerializationContext(
+                                                                           fan_out_message.topic(),
+                                                                           MessageField.VALUE))
+                _log.info("Unpacked message as %r.", deserialized_fan_out_visit)
+
+                # Calculate time to load knative and receive message based on timestamp in Kafka message
+                _log.debug("Message timestamp %r", fan_out_message.timestamp())
+                fan_out_kafka_msg_timestamp = fan_out_message.timestamp()
+                fan_out_to_prompt_time = int(time.time() * 1000) - fan_out_kafka_msg_timestamp[1]
+                _log.debug("Seconds since fan out message delivered %r", fan_out_to_prompt_time/1000)
+
+                # Commit message and close client
+                fan_out_consumer.commit(message=fan_out_message, asynchronous=False)
+                fan_out_consumer.close()
+
+                # Process fan out visit
+                process_visit(deserialized_fan_out_visit)
+                _log.info("Processing completed for %s", socket.gethostname())
```

Comment:
I recommend putting this in a finally block to be analogous to [the Knative message](https://github.com/lsst-dm/prompt_processing/blob/6f3c9f86eb420847ee8b65a7d846d6d1f1bd4957/python/activator/activator.py#L531). `process_visit` raises exceptions fairly often.